### PR TITLE
Prometheus: Support request cancellation properly (Uses new backendSrv.fetch Observable request API) 

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
This PR replaces the usage of datasourceRequest in the Prometheus data source with fetch and updates tests accordingly.

**Which issue(s) this PR fixes**:
Fixes #

Relates #27222

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/datasource.ts
Comment (hugohaggmark) [inline public/app/plugins/datasource/prometheus/datasource.ts:700]: Changed to `any` here because of `const body = res.data || res.json();` complaining of `json` not being defined. I don't understand how `res.json()` can work as this is already done in backendSrv. Might be something old that I can remove? WDYT @grafana/observability-squad

File: public/app/plugins/datasource/prometheus/language_provider.ts
Comment (hugohaggmark) [inline public/app/plugins/datasource/prometheus/language_provider.ts:103]: Same comment as above :)

File: public/app/plugins/datasource/prometheus/datasource.ts
Comment (davkal) [reply public/app/plugins/datasource/prometheus/datasource.ts:700]: I think this was done for testing where `data` would be the response that the test injected, but this might have evolved to no longer being used.

Comment (hugohaggmark) [reply public/app/plugins/datasource/prometheus/datasource.ts:700]: Thank you David, I think I'll remove it and see if it explodes in the wild💥

**Review by dprokop — APPROVED** [review]
This looks good @hugohaggmark!

Just a thought about `getBackendSrv().fetch<T>(options)` vs  `getBackendSrv().datasourceRequest(options)`. `fetch` feels kind of low level function to me(name-wise). Especially that for GET, POST etc requests we have wrappers around it. Doesn't really matter they return a promise, it's just about a convention. After a few weeks after fetch was introduced and reading this PR I  think `datasourceRequest` is/was a good name do use in the data source code, as the API was explicit for the data source context. Even if the method implementation was as dummy as :

```
datasourceRequest(options: BackendSrvRequest):  Observable<FetchResponse<T>> {
    return this.fetch(options);
 }
```

then discoverabilty-wise the API feels a little bit more friendly:) Yeah, better late than never :D

**Comment by hugohaggmark** [comment]
Waiting for an approval from the observability squad 🙏

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:23]: Nit: Typo
```suggestion
const fetchMock = jest.spyOn(backendSrv, 'fetch');
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:61]: Same as above
```suggestion
    fetchMock.mockImplementation(() => of(({ status: 'success', data: data.response } as unknown) as FetchResponse));
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:77]: Same as above
```suggestion
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith({
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:96]: Same as above
```suggestion
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith({
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:119]:
```suggestion
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith({
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:144]:
```suggestion
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith({
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:170]:
```suggestion
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith({
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:191]:
```suggestion
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith({
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:219]:
```suggestion
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith({
```

File: public/app/plugins/datasource/prometheus/metric_find_query.test.ts
Comment (kaydelaney) [inline public/app/plugins/datasource/prometheus/metric_find_query.test.ts:245]:
```suggestion
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith({
```

**Review by kaydelaney — APPROVED** [review]
Just some comments about a typo but otherwise LGTM!

**Comment by hugohaggmark** [comment]
> Just some comments about a typo but otherwise LGTM!

Thank you @kaydelaney ⭐

**Review by ivanahuckova — APPROVED** [review]
Joining the LGTM 🚂 . Thank you so much for doing this @hugohaggmark!

**Pull request merged**
